pkg/nodeagent/controller/operatingsystemconfig/containerd.go

```diff
@@ -0,0 +1,363 @@
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		return nil
+	}
+
+	output, err := Exec(ctx, "containerd", "config", "default")
+	if err != nil {
+		return err
+	}
+
+	return r.FS.WriteFile(configFile, output, 0644)
+}
+
+// ensureContainerdEnvironment sets the environment for the 'containerd' service.
+func (r *Reconciler) ensureContainerdEnvironment() error {
+	var (
+		unitDropin = `[Service]
+Environment="PATH=` + extensionsv1alpha1.ContainerDRuntimeContainersBinFolder + `:` + os.Getenv("PATH") + `"
+`
+	)
+
+	containerdEnvFilePath := path.Join(dropinDir, "30-env_config.conf")
+	exists, err := r.fileExists(containerdEnvFilePath)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		return nil
+	}
+
+	err = r.FS.WriteFile(containerdEnvFilePath, []byte(unitDropin), 0644)
+	if err != nil {
+		return fmt.Errorf("unable to write unit dropin: %w", err)
+	}
+
+	return nil
+}
+
+// ensureContainerdConfiguration sets the configuration for containerd.
+func (r *Reconciler) ensureContainerdConfiguration(criConfig *extensionsv1alpha1.CRIConfig) error {
+	config, err := r.FS.ReadFile(configFile)
+	if err != nil {
+		return fmt.Errorf("unable to read containerd config.toml: %w", err)
+	}
+
+	content := map[string]any{}
+
+	if err = toml.Unmarshal(config, &content); err != nil {
+		return fmt.Errorf("unable to decode containerd default config: %w", err)
+	}
+
+	type (
+		patch struct {
+			name  string
+			path  structuredmap.Path
+			setFn structuredmap.SetFn
+		}
+	)
+
+	patches := []patch{
+		{
+			name: "registry config path",
+			path: structuredmap.Path{"plugins", "io.containerd.grpc.v1.cri", "registry", "config_path"},
+			setFn: func(_ any) (any, error) {
+				return certsDir, nil
+			},
+		},
+		{
+			name: "imports paths",
+			path: structuredmap.Path{"imports"},
+			setFn: func(value any) (any, error) {
+				importPath := path.Join(configDir, "*.toml")
+
+				imports, ok := value.([]any)
+				if !ok {
+					return []string{importPath}, nil
+				}
+
+				for _, imp := range imports {
+					path, ok := imp.(string)
+					if !ok {
+						continue
+					}
+
+					if path == importPath {
+						return value, nil
+					}
+				}
+
+				return append(imports, importPath), nil
+			},
+		},
+		{
+			name: "sandbox image",
+			path: structuredmap.Path{"plugins", "io.containerd.grpc.v1.cri", "sandbox_image"},
+			setFn: func(value any) (any, error) {
+				if criConfig == nil || criConfig.Containerd == nil {
+					return value, nil
+				}
+
+				return criConfig.Containerd.SandboxImage, nil
+			},
+		},
+	}
+
+	for _, p := range patches {
+		content, err = structuredmap.SetMapEntry(content, p.path, p.setFn)
+		if err != nil {
+			return fmt.Errorf("unable setting %q in containerd config.toml: %w", p.name, err)
+		}
+	}
+
+	f, err := r.FS.OpenFile(configFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("unable to open containerd config.toml: %w", err)
+	}
+	defer func() {
+		err = f.Close()
+	}()
```

Comment:
If err != nil, err won't be returned to the caller. Not sure if this is expected.